File: generate_import_v2.py
"""
generate_import_v2.py
Reads the SharePoint CSV export and produces import_investment_ideas_v2.sql.
Run: python generate_import_v2.py
"""

import csv, uuid, re, json

CSV_PATH    = r"C:\Users\chelseako\Downloads\2026 USP Data & AI Pipeline Planning List (4).csv"
OUTPUT_PATH = r"C:\Users\chelseako\usp-dataai-insights-hub\import_investment_ideas_v2.sql"
SCHEMA      = "usp_data.usp_strategy"

# ── Lookup tables ─────────────────────────────────────────────────────────────

PORTFOLIO_MAP = {
    "AI Infrastructure":        "ai-infra",
    "System Feedback Loops":    "sfl",
    "Data & AI Enablement Hub": "hub",
    "Cross-cutting Supports":   "cross-cutting",
    "Reserve":                  "reserve",
}

BOW_MAP = {
    # New names
    "Enhance Context & Personalization":                  "ai-infra-bow1",
    "Accelerate AI Evaluation, Evidence, and Guardrails": "ai-infra-bow2",
    "Mobilize Frontier Labs for Learner Success":         "ai-infra-bow3",
    "Build and Sustain EDU-Net":                          "sfl-bow1",
    "Advance Data & AI Feedback Loops in Place":          "sfl-bow2",
    "Launch Competencies & Skills Genome Accelerator":    "sfl-bow3",
    "Launch Competencies and Skills Genome Accelerator":  "sfl-bow3",
    "Enable Data & AI Capabilities for Division":         "hub-bow1",
    "Advance Strategy Learning & Insight":                "bow1",
    "Enable Business & Strategy Execution":               "bow2",
    # Old / variant names kept for safety
    "Build Model Capabilities & Learner Context":         "ai-infra-bow1",
    "Develop AI Evaluation Infrastructure":               "ai-infra-bow2",
    "Integrated Delivery":                                "sfl-bow2",
    "AI Evaluation Technologies & Evidence Acceleration": "ai-infra-bow2",
    "Portable Memory & Context":                          "ai-infra-bow1",
    "Hyperscaler Consortium":                             "ai-infra-bow3",
    "EDU-Net":                                            "sfl-bow1",
    "Data in Place":                                      "sfl-bow2",
    "Competencies & Skills Genome Accelerator":           "sfl-bow3",
    "Cross-division AI & Data Support and Learning":      "hub-bow1",
    "Insights for AI in Education":                       "hub-bow1",
    "MLE / Impact Accounting":                            "bow1",
    "Business and Strategy Support":                      "bow2",
}

STAGE_MAP = {
    "OK to Proceed":   "Okay to Proceed",
    "Moved to INVEST": "Moved to Invest",
    # pass-through
    "Brainstorming":    "Brainstorming",
    "Ready for Review": "Ready for Review",
    "More Info Needed": "More Info Needed",
    "On Hold":          "On Hold",
    "Moved to Invest":  "Moved to Invest",
    "Okay to Proceed":  "Okay to Proceed",
}

EMAIL_MAP = {
    "brianna.moore-trieu@gatesfoundation.org": "Brianna Moore-Trieu",
    "chelsea.koproske@gatesfoundation.org":    "Chelsea Koproske",
    "julia.bloomweltman@gatesfoundation.org":  "Julia Bloom-Weltman",
    "lilian.tan@gatesfoundation.org":          "Lilian Tan",
    "marten.roorda@gatesfoundation.org":       "Marten Roorda",
    "matt.gee@gatesfoundation.org":            "Matt Gee",
    "megan.coolidge@gatesfoundation.org":      "Megan Coolidge",
    "nicole.ifill@gatesfoundation.org":        "Nicole Ifill",
    "pavani.reddy@gatesfoundation.org":        "Pavani Reddy",
    "rachel.scherer@gatesfoundation.org":      "Rachel Scherer",
    "vic.vuchic@gatesfoundation.org":          "Vic Vuchic",
    "xiaoxue.du@gatesfoundation.org":          "Xiaoxue Du",
}

# ── Helpers ───────────────────────────────────────────────────────────────────

def esc(s):
    """Escape a string for SQL single-quoted literal, or return NULL."""
    if s is None or str(s).strip() == "":
        return "NULL"
    return "'" + str(s).replace("'", "''") + "'"

def parse_amount(s):
    if not s or str(s).strip() in ("", "0", "$0"):
        return "NULL"
    s = re.sub(r"[$,\s]", "", str(s).strip())
    try:
        v = float(s)
        return str(int(v)) if v == int(v) else str(v)
    except Exception:
        return "NULL"

def resolve_email(raw):
    if not raw or not raw.strip():
        return None
    return EMAIL_MAP.get(raw.strip().lower()) or EMAIL_MAP.get(raw.strip()) or raw.strip()

def resolve_bow(name):
    if not name or not name.strip():
        return None
    return BOW_MAP.get(name.strip())

def resolve_portfolio(name):
    if not name or not name.strip():
        return None
    return PORTFOLIO_MAP.get(name.strip())

def resolve_stage(raw):
    if not raw or not raw.strip():
        return "Brainstorming"
    return STAGE_MAP.get(raw.strip(), raw.strip())

def parse_additional_bows(raw):
    """Parse the Additional BOW(s) JSON array field from SharePoint."""
    if not raw or not raw.strip():
        return None
    try:
        # SharePoint exports multi-choice as JSON array string
        items = json.loads(raw.strip())
        bow_ids = [BOW_MAP.get(i.strip()) for i in items if i.strip()]
        bow_ids = [b for b in bow_ids if b]
        return ";".join(bow_ids) if bow_ids else None
    except Exception:
        # Fallback: try as plain text
        mapped = BOW_MAP.get(raw.strip())
        return mapped

# ── Read CSV ──────────────────────────────────────────────────────────────────
# Row 0: SharePoint schema XML  → skip
# Row 1: column headers
# Rows 2+: data

data_rows = []
with open(CSV_PATH, newline="", encoding="utf-8-sig") as fh:
    reader = csv.reader(fh)
    next(reader)          # skip schema XML row
    headers = next(reader)
    for row in reader:
        if not any(c.strip() for c in row):
            continue
        while len(row) < len(headers):
            row.append("")
        data_rows.append(dict(zip(headers, row)))

print(f"Parsed {len(data_rows)} rows from CSV")

# ── Build VALUES ──────────────────────────────────────────────────────────────

unmapped_bows = set()
unmapped_portfolios = set()
value_parts = []

for r in data_rows:
    title = r.get("Investment Idea", "").strip()
    if not title:
        continue

    idea_id      = str(uuid.uuid4())
    submitted_by = resolve_email(r.get("Submitted By", ""))
    inv_raw      = r.get("INV Number", "").strip() or None
    stage        = resolve_stage(r.get("Stage", ""))
    objective    = r.get("Objective / Intended Impact", "").strip() or None
    idea_type    = r.get("Idea Type", "").strip() or None

    port_raw     = r.get("Primary Portfolio", "").strip()
    port_id      = resolve_portfolio(port_raw)
    if port_raw and not port_id:
        unmapped_portfolios.add(port_raw)

    bow_raw      = r.get("Primary Body of Work", "").strip()
    bow_id       = resolve_bow(bow_raw)
    if bow_raw and not bow_id:
        unmapped_bows.add(bow_raw)

    add_bows_raw = r.get("Additional BOW(s)", "").strip()
    add_bows     = parse_additional_bows(add_bows_raw)

    partner      = r.get("Potential Partner", "").strip() or None
    est_total    = parse_amount(r.get("Est. Total Data $", ""))
    est_2026     = parse_amount(r.get("Est. Data $ in 2026", ""))
    co_fund      = r.get("Potential Co-funding Details", "").strip() or None
    start_date   = r.get("Desired Start Date", "").strip() or None
    duration     = r.get("Est. Duration", "").strip() or None
    notes        = r.get("Additional Notes", "").strip() or None
    appr_note    = r.get("Approver Comments (Optional)", "").strip() or None

    moved        = (stage == "Moved to Invest")

    v = (
        f"  ({esc(idea_id)}, {esc(title)}, {esc(submitted_by)}, current_timestamp(),\n"
        f"   {esc(stage)}, {esc(idea_type)}, {esc(objective)},\n"
        f"   {esc(port_id)}, {esc(bow_id)}, {esc(add_bows)},\n"
        f"   {esc(partner)}, {est_total}, {est_2026},\n"
        f"   {esc(co_fund)}, {esc(start_date)}, {esc(duration)},\n"
        f"   {esc(notes)}, {esc(appr_note)},\n"
        f"   {esc(inv_raw)}, {'true' if moved else 'false'}, false)"
    )
    value_parts.append(v)

# ── Write SQL ─────────────────────────────────────────────────────────────────

sql = f"""\
-- ============================================================
-- Investment Ideas re-import ({len(value_parts)} records)
-- Source: 2026 USP Data & AI Pipeline Planning List (4).csv
-- Generated by generate_import_v2.py
--
-- Run both statements in a single Databricks SQL session.
-- The DELETE clears non-archived ideas; INSERT adds all {len(value_parts)}.
-- ============================================================

-- Step 1: Clear existing active ideas
DELETE FROM {SCHEMA}.investment_ideas
WHERE COALESCE(archived, false) = false;

-- Step 2: Insert all {len(value_parts)} ideas
INSERT INTO {SCHEMA}.investment_ideas
    (idea_id, title, submitted_by, submitted_at,
     stage, idea_type, objective,
     primary_portfolio, primary_bow, additional_bows,
     potential_partner, est_total_amount, est_2026_amount,
     co_funding_details, desired_start_date, est_duration,
     notes, approver_note,
     inv_number, moved_to_invest, archived)
VALUES
{",\n".join(value_parts)};
"""

with open(OUTPUT_PATH, "w", encoding="utf-8") as fh:
    fh.write(sql)

print(f"\nWrote {OUTPUT_PATH}")
print(f"  Rows:  {len(value_parts)}")

if unmapped_bows:
    print(f"\n⚠ Unmapped BOW names (stored as NULL):")
    for b in sorted(unmapped_bows):
        print(f"    {b!r}")

if unmapped_portfolios:
    print(f"\n⚠ Unmapped portfolio names (stored as NULL):")
    for p in sorted(unmapped_portfolios):
        print(f"    {p!r}")

File: generate_import_v2.js
/**
 * generate_import_v2.js
 * Reads the SharePoint CSV and produces import_investment_ideas_v2.sql
 * Run: node generate_import_v2.js
 */

const fs   = require("fs");
const path = require("path");
const { randomUUID } = require("crypto");

const CSV_PATH    = String.raw`C:\Users\chelseako\Downloads\2026 USP Data & AI Pipeline Planning List (4).csv`;
const OUTPUT_PATH = path.join(__dirname, "import_investment_ideas_v2.sql");
const SCHEMA      = "usp_data.usp_strategy";

// ── Lookup tables ─────────────────────────────────────────────────────────────

const PORTFOLIO_MAP = {
  "AI Infrastructure":        "ai-infra",
  "System Feedback Loops":    "sfl",
  "Data & AI Enablement Hub": "hub",
  "Cross-cutting Supports":   "cross-cutting",
  "Reserve":                  "reserve",
};

const BOW_MAP = {
  // Current names
  "Enhance Context & Personalization":                  "ai-infra-bow1",
  "Accelerate AI Evaluation, Evidence, and Guardrails": "ai-infra-bow2",
  "Mobilize Frontier Labs for Learner Success":         "ai-infra-bow3",
  "Build and Sustain EDU-Net":                          "sfl-bow1",
  "Advance Data & AI Feedback Loops in Place":          "sfl-bow2",
  "Launch Competencies & Skills Genome Accelerator":    "sfl-bow3",
  "Launch Competencies and Skills Genome Accelerator":  "sfl-bow3",
  "Enable Data & AI Capabilities for Division":         "hub-bow1",
  "Advance Strategy Learning & Insight":                "bow1",
  "Enable Business & Strategy Execution":               "bow2",
  // Historic / variant names
  "Build Model Capabilities & Learner Context":         "ai-infra-bow1",
  "Develop AI Evaluation Infrastructure":               "ai-infra-bow2",
  "Integrated Delivery":                                "sfl-bow2",
  "AI Evaluation Technologies & Evidence Acceleration": "ai-infra-bow2",
  "Portable Memory & Context":                          "ai-infra-bow1",
  "Hyperscaler Consortium":                             "ai-infra-bow3",
  "EDU-Net":                                            "sfl-bow1",
  "Data in Place":                                      "sfl-bow2",
  "Competencies & Skills Genome Accelerator":           "sfl-bow3",
  "Cross-division AI & Data Support and Learning":      "hub-bow1",
  "Insights for AI in Education":                       "hub-bow1",
  "MLE / Impact Accounting":                            "bow1",
  "Business and Strategy Support":                      "bow2",
};

const STAGE_MAP = {
  "OK to Proceed":   "Okay to Proceed",
  "Moved to INVEST": "Moved to Invest",
  "Brainstorming":   "Brainstorming",
  "Ready for Review":"Ready for Review",
  "More Info Needed":"More Info Needed",
  "On Hold":         "On Hold",
  "Moved to Invest": "Moved to Invest",
  "Okay to Proceed": "Okay to Proceed",
};

const EMAIL_MAP = {
  "brianna.moore-trieu@gatesfoundation.org": "Brianna Moore-Trieu",
  "chelsea.koproske@gatesfoundation.org":    "Chelsea Koproske",
  "julia.bloomweltman@gatesfoundation.org":  "Julia Bloom-Weltman",
  "lilian.tan@gatesfoundation.org":          "Lilian Tan",
  "marten.roorda@gatesfoundation.org":       "Marten Roorda",
  "matt.gee@gatesfoundation.org":            "Matt Gee",
  "megan.coolidge@gatesfoundation.org":      "Megan Coolidge",
  "nicole.ifill@gatesfoundation.org":        "Nicole Ifill",
  "pavani.reddy@gatesfoundation.org":        "Pavani Reddy",
  "rachel.scherer@gatesfoundation.org":      "Rachel Scherer",
  "vic.vuchic@gatesfoundation.org":          "Vic Vuchic",
  "xiaoxue.du@gatesfoundation.org":          "Xiaoxue Du",
};

// ── Helpers ───────────────────────────────────────────────────────────────────

function esc(s) {
  if (s === null || s === undefined || String(s).trim() === "") return "NULL";
  return "'" + String(s).replace(/'/g, "''") + "'";
}

function parseAmount(s) {
  if (!s || !String(s).trim()) return "NULL";
  const cleaned = String(s).replace(/[$,\s]/g, "").trim();
  if (!cleaned || cleaned === "0") return "NULL";
  const n = parseFloat(cleaned);
  if (isNaN(n)) return "NULL";
  return String(Number.isInteger(n) ? n : n);
}

function resolveEmail(raw) {
  if (!raw || !raw.trim()) return null;
  return EMAIL_MAP[raw.trim().toLowerCase()] || EMAIL_MAP[raw.trim()] || raw.trim();
}

function normStr(s) {
  // Normalize: trim, collapse whitespace, decode HTML entities
  return s.trim()
    .replace(/&amp;/g, "&").replace(/&#38;/g, "&")
    .replace(/&quot;/g, '"').replace(/&#34;/g, '"')
    .replace(/\s+/g, " ");
}

function resolveBow(name) {
  if (!name || !name.trim()) return null;
  return BOW_MAP[normStr(name)] || BOW_MAP[name.trim()] || null;
}

function resolvePortfolio(name) {
  if (!name || !name.trim()) return null;
  return PORTFOLIO_MAP[name.trim()] || null;
}

function resolveStage(raw) {
  if (!raw || !raw.trim()) return "Brainstorming";
  return STAGE_MAP[raw.trim()] || raw.trim();
}

function parseAdditionalBows(raw) {
  if (!raw || !raw.trim()) return null;
  try {
    const items = JSON.parse(raw.trim());
    if (Array.isArray(items)) {
      const ids = items.map(i => BOW_MAP[i.trim()]).filter(Boolean);
      return ids.length ? ids.join(";") : null;
    }
  } catch (e) { /* not JSON */ }
  const mapped = BOW_MAP[raw.trim()];
  return mapped || null;
}

// ── CSV parser (handles quoted multi-line fields) ─────────────────────────────

function parseCSV(text) {
  const rows = [];
  let cur = [], field = "", inQuote = false, i = 0;
  while (i < text.length) {
    const ch = text[i];
    if (inQuote) {
      if (ch === '"') {
        if (text[i+1] === '"') { field += '"'; i += 2; continue; }
        inQuote = false; i++; continue;
      }
      field += ch; i++; continue;
    }
    if (ch === '"') { inQuote = true; i++; continue; }
    if (ch === ',') { cur.push(field); field = ""; i++; continue; }
    if (ch === '\r' && text[i+1] === '\n') {
      cur.push(field); field = ""; rows.push(cur); cur = []; i += 2; continue;
    }
    if (ch === '\n') {
      cur.push(field); field = ""; rows.push(cur); cur = []; i++; continue;
    }
    field += ch; i++;
  }
  if (field || cur.length) { cur.push(field); rows.push(cur); }
  return rows;
}

// ── Main ──────────────────────────────────────────────────────────────────────

const raw   = fs.readFileSync(CSV_PATH, "utf-8");
const allRows = parseCSV(raw);

// Row 0 = schema XML, Row 1 = headers, Rows 2+ = data
const headers  = allRows[1];
const dataRows = allRows.slice(2).filter(r => r.some(c => c.trim()));

console.log(`Parsed ${dataRows.length} data rows`);

const unmappedBows  = new Set();
const unmappedPorts = new Set();
const valueParts    = [];

for (const row of dataRows) {
  while (row.length < headers.length) row.push("");
  const r = Object.fromEntries(headers.map((h, i) => [h, row[i] || ""]));

  const title = r["Investment Idea"].trim();
  if (!title) continue;

  const idea_id      = randomUUID();
  const submitted_by = resolveEmail(r["Submitted By"]);
  const inv_raw      = r["INV Number"].trim() || null;
  const stage        = resolveStage(r["Stage"]);
  const objective    = r["Objective / Intended Impact"].trim() || null;
  const idea_type    = r["Idea Type"].trim() || null;

  const port_raw = r["Primary Portfolio"].trim();
  const port_id  = resolvePortfolio(port_raw);
  if (port_raw && !port_id) unmappedPorts.add(port_raw);

  const bow_raw  = r["Primary Body of Work"].trim();
  const bow_id   = resolveBow(bow_raw);
  if (bow_raw && !bow_id) {
    unmappedBows.add(bow_raw);
    console.log(`  ⚠ Row "${title.substring(0,40)}": bow_raw=${JSON.stringify(bow_raw)}`);
  }

  const add_bows = parseAdditionalBows(r["Additional BOW(s)"]);
  const partner  = r["Potential Partner"].trim() || null;
  const est_total= parseAmount(r["Est. Total Data $"]);
  const est_2026 = parseAmount(r["Est. Data $ in 2026"]);
  const co_fund  = r["Potential Co-funding Details"].trim() || null;
  const start_dt = r["Desired Start Date"].trim() || null;
  const duration = r["Est. Duration"].trim() || null;
  const notes    = r["Additional Notes"].trim() || null;
  const appr_note= r["Approver Comments (Optional)"].trim() || null;

  const moved = (stage === "Moved to Invest");

  valueParts.push(
    `  (${esc(idea_id)}, ${esc(title)}, ${esc(submitted_by)}, current_timestamp(),\n` +
    `   ${esc(stage)}, ${esc(idea_type)}, ${esc(objective)},\n` +
    `   ${esc(port_id)}, ${esc(bow_id)}, ${esc(add_bows)},\n` +
    `   ${esc(partner)}, ${est_total}, ${est_2026},\n` +
    `   ${esc(co_fund)}, ${esc(start_dt)}, ${esc(duration)},\n` +
    `   ${esc(notes)}, ${esc(appr_note)},\n` +
    `   ${esc(inv_raw)}, ${moved ? "true" : "false"}, false)`
  );
}

const sql =
`-- ============================================================
-- Investment Ideas re-import (${valueParts.length} records)
-- Source: 2026 USP Data & AI Pipeline Planning List (4).csv
-- Generated by generate_import_v2.js
--
-- Run BOTH statements in a single Databricks SQL session.
-- Step 1 clears non-archived ideas; Step 2 inserts all ${valueParts.length}.
-- ============================================================

-- Step 1: Clear existing active (non-archived) ideas
DELETE FROM ${SCHEMA}.investment_ideas
WHERE COALESCE(archived, false) = false;

-- Step 2: Insert all ${valueParts.length} ideas from SharePoint export
INSERT INTO ${SCHEMA}.investment_ideas
    (idea_id, title, submitted_by, submitted_at,
     stage, idea_type, objective,
     primary_portfolio, primary_bow, additional_bows,
     potential_partner, est_total_amount, est_2026_amount,
     co_funding_details, desired_start_date, est_duration,
     notes, approver_note,
     inv_number, moved_to_invest, archived)
VALUES
${valueParts.join(",\n")};
`;

fs.writeFileSync(OUTPUT_PATH, sql, "utf-8");

console.log(`\nWrote ${OUTPUT_PATH}`);
console.log(`  Rows: ${valueParts.length}`);

if (unmappedBows.size) {
  console.log(`\n⚠ Unmapped BOW names (stored as NULL):`);
  for (const b of [...unmappedBows].sort()) console.log(`    "${b}"`);
}
if (unmappedPorts.size) {
  console.log(`\n⚠ Unmapped portfolio names (stored as NULL):`);
  for (const p of [...unmappedPorts].sort()) console.log(`    "${p}"`);
}
